datastore/datastore.go

```diff
@@ -0,0 +1,217 @@
+package datastore
+
+import (
+	"bytes"
+	"encoding/gob"
+	"errors"
+	"reflect"
+	"strings"
+
+	"github.com/docker/libnetwork/config"
+	"github.com/docker/swarm/pkg/store"
+)
+
+//DataStore exported
+type DataStore interface {
+	// PutObject adds a new Record based on an object into the datastore
+	PutObject(kvObject KV) error
+	// PutObjectAtomic provides an atomic add and update operation for a Record
+	PutObjectAtomic(kvObject KV) error
+	// KVStore returns access to the KV Store
+	KVStore() store.Store
+}
+
+type datastore struct {
+	store store.Store
+}
+
+//KV Key Value interface used by objects to be part of the DataStore
+type KV interface {
+	Key() []string
+	Value() []byte
+	Index() uint64
+	SetIndex(uint64)
+}
+
+//Key provides convenient method to create a Key
+func Key(key ...string) string {
+	keychain := []string{"docker", "libnetwork"}
+	keychain = append(keychain, key...)
+	str := strings.Join(keychain, "/")
+	return str + "/"
+}
+
+var errNewDatastore = errors.New("Error creating new Datastore")
+var errInvalidConfiguration = errors.New("Invalid Configuration passed to Datastore")
+
+// newClient used to connect to KV Store
+func newClient(kv string, addrs string) (DataStore, error) {
+	store, err := store.CreateStore(kv, []string{addrs}, store.Config{})
+	if err != nil {
+		return nil, err
+	}
+	ds := &datastore{store: store}
+	return ds, nil
+}
+
+// NewDataStore creates a new instance of LibKV data store
+func NewDataStore(cfg *config.DatastoreCfg) (DataStore, error) {
+	if cfg == nil {
+		return nil, errInvalidConfiguration
+	}
+	// TODO : cfg.Embedded case
+	return newClient(cfg.Client.Provider, cfg.Client.Address)
+}
+
+// NewCustomDataStore can be used by clients to plugin cusom datatore that adhers to store.Store
+func NewCustomDataStore(customStore store.Store) DataStore {
+	return &datastore{store: customStore}
+}
+
+func (ds *datastore) KVStore() store.Store {
+	return ds.store
+}
+
+// PutObjectAtomic adds a new Record based on an object into the datastore
+func (ds *datastore) PutObjectAtomic(kvObject KV) error {
+	if kvObject == nil {
+		return errors.New("kvObject is nil")
+	}
+	kvObjValue := kvObject.Value()
+
+	if kvObjValue == nil {
+		return ErrInvalidAtomicRequest
+	}
+	_, err := ds.store.AtomicPut(Key(kvObject.Key()...), []byte{}, kvObjValue, kvObject.Index())
```

Comment:
Why are you passing an empty `[]byte{}` for value? From what I understand(from the consul implementation of the KV store) this seems to the value that is going to be set at that key and the 3rd argument is supposed to be the previous value. So I am not sure how this was working?